app/app/src/main/java/com/example/myapplication/MainActivity.java

```diff
@@ -410,6 +423,52 @@ void goBackward() throws IOException {
         int msg = 5;
         Car.mmOutputStream.write(msg);
     }
+
+    // joyStick control methods to replace four buttons
+    public void onMove(JoyStick joyStick, double angle, double power, int direction) {
+        switch (direction){
+            case JoyStick.DIRECTION_LEFT:
+                Toast.makeText(MainActivity.this,"Go Left",Toast.LENGTH_SHORT).show();
```

Comment:
`Toast.LENGTH_SHORT` is not short enough for a toast to disappear before a new toast is coming, if you move the joystick quickly. This can be fixed with a custom duration, but I think it's also safe to say that the end-user understands that they're going forward if they move forward.